### PR TITLE
fix: AVRO-4055: Validate records properly when parsing a schema

### PR DESCRIPTION
**Pull request by woile**
See https://issues.apache.org/jira/browse/AVRO-4055

### CODE REVIEW

File: avro/src/error.rs [hunk 1]
```diff
@@ -350,6 +350,9 @@ pub enum Error {
     #[error("Invalid namespace {0}. It must match the regex '{1}'")]
     InvalidNamespace(String, &'static str),
 
+    #[error("Invalid schema: The record type should be inside the type field")]
+    InvalidSchemaRecord(String),
```
Comment (martin-g) [inline avro/src/error.rs:354]: The argument is not printed in the error message, i.e. there is no `{0}`.

File: avro/src/schema.rs [hunk 2]
```diff
@@ -1353,6 +1358,7 @@ impl Parser {
         &mut self,
         complex: &Map<String, Value>,
         enclosing_namespace: &Namespace,
+        from_field: bool,
```
Comment (martin-g) [inline avro/src/schema.rs]: IMO it would be better to use an `enum` with well named variants instead of `true` / `false`

File: avro/src/schema.rs [hunk 3]
```diff
@@ -6821,4 +6832,63 @@ mod tests {
...
+                "default": null
+                },
+                { "name": "order_date", "type": "string" }
+            ]
+            }
+        ]
+        }"#;
+
+        let schema = Schema::parse_str(invalid_schema_str);
+        println!("{:?}", schema);
```
Comment (martin-g) [inline avro/src/schema.rs]: debug leftover

File: avro/src/types.rs [hunk 4]
```diff
@@ -2771,35 +2771,36 @@ Field with name '"b"' is not a member of the map items"#,
         use crate::ser::Serializer;
         use serde::Serialize;
 
-        let schema_str = r#"{
+        let schema_str = r#"
+        {
             "type": "record",
             "name": "NamespacedMessage",
-            [NAMESPACE]
```
Comment (martin-g) [inline avro/src/types.rs:2777]: Why is this removed ?

Comment (woile) [reply avro/src/types.rs:2777]: Oh my bad, it took me some time to fix this schema, and then I forgot to add the namespace again, I'll update it

File: avro/src/schema.rs (hunk 2 above)
Comment (woile) [reply avro/src/schema.rs]: Great idea! I'm updating it

**Comment by woile** [comment]
Updated based on the feedback!

File: avro/src/schema.rs [hunk 5]
```diff
@@ -1294,6 +1307,11 @@ impl Parser {
             return Ok(resolving_schema.clone());
         }
 
+        // For good error reporting we add this check
+        if name.name == *"record" {
```
Comment (martin-g) [inline avro/src/schema.rs]: I am not sure this is always valid.
`#fetch_schema_ref()` is called also by `#parse_known_schema()` where the `name` argument could be any non-primitive schema kind, like `"MyType"` (a Schema::Ref).

Comment (martin-g) [reply avro/src/schema.rs]: I misunderstood you idea. Let me think how to improve it!

Comment (woile) [reply avro/src/schema.rs]: In a nutshell, if you have defined a type called "record" somewhere else, then you can actually do `"type": "record"`, but if there's no schema already defined, then you probably intended to create a new record, so it should be inside type, as `"type": {"type": "record"}`. Does it make sense?

**Review by martin-g — APPROVED** [review]

**Comment by martin-g** [comment]
Thank you, @woile !
